docs/howto/csrf-protection.txt

```diff
@@ -0,0 +1,374 @@
+response will be cached without either one. In this case, on any views that
+will require a CSRF token to be inserted you should use the
+:func:`django.views.decorators.csrf.csrf_protect` decorator first::
+
+  from django.views.decorators.cache import cache_page
+  from django.views.decorators.csrf import csrf_protect
+
+  @cache_page(60 * 15)
+  @csrf_protect
+  def my_view(request):
+      ...
+
+If you are using class-based views, you can refer to :ref:`Decorating
+class-based views<decorating-class-based-views>`.
+
+Testing and CSRF protection
+===========================
+
+The ``CsrfViewMiddleware`` will usually be a big hindrance to testing view
+functions, due to the need for the CSRF token which must be sent with every POST
+request. For this reason, Django's HTTP client for tests has been modified to
+set a flag on requests which relaxes the middleware and the ``csrf_protect``
+decorator so that they no longer rejects requests. In every other respect
+(e.g. sending cookies etc.), they behave the same.
+
+If, for some reason, you *want* the test client to perform CSRF
+checks, you can create an instance of the test client that enforces
+CSRF checks::
+
+    >>> from django.test import Client
+    >>> csrf_client = Client(enforce_csrf_checks=True)
+
+Edge cases
+==========
+
+Certain views can have unusual requirements that mean they don't fit the normal
+pattern envisaged here. A number of utilities can be useful in these
+situations. The scenarios they might be needed in are described in the following
+section.
+
+Utilities
+---------
+
+The examples below assume you are using function-based views. If you
+are working with class-based views, you can refer to :ref:`Decorating
+class-based views<decorating-class-based-views>`.
+
+.. function:: csrf_exempt(view)
+
+    This decorator marks a view as being exempt from the protection ensured by
+    the middleware. Example::
+
+        from django.http import HttpResponse
+        from django.views.decorators.csrf import csrf_exempt
+
+        @csrf_exempt
+        def my_view(request):
+            return HttpResponse('Hello world')
+
+.. function:: requires_csrf_token(view)
+
+    Normally the :ttag:`csrf_token` template tag will not work if
+    ``CsrfViewMiddleware.process_view`` or an equivalent like ``csrf_protect``
+    has not run. The view decorator ``requires_csrf_token`` can be used to
+    ensure the template tag does work. This decorator works similarly to
+    ``csrf_protect``, but never rejects an incoming request.
+
+    Example::
+
+        from django.shortcuts import render
+        from django.views.decorators.csrf import requires_csrf_token
+
+        @requires_csrf_token
+        def my_view(request):
+            c = {}
+            # ...
+            return render(request, "a_template.html", c)
+
+.. function:: ensure_csrf_cookie(view)
+
+    This decorator forces a view to send the CSRF cookie.
+
```

Comment:
This should stay in `ref`.

Reply:
Fixed.
I added `:noindex:` in `howto/csrf` because the documentation for `django.views.decorators.csrf` is now split in both files (`csrf_protect` is documented in `howto/csrf`, `csrf_exempt`, `requires_csrf_token` and `ensure_csrf_cookie` are documented in `ref/csrf` under "Utilities")
 Feel free to comment or move it if this is a problem or should be moved elsewhere. If not you can go ahead and resolve this.

Reply:
Thanks :+1: I moved `csrf_protect` docs to the `ref`.